rBasex, a better pBasex

MikhailRyazanov opened the pull request:
Here is a new method for analyzing velocity-mapping images, similar to pBasex, but implemented more efficiently.

The main idea is that the Abel transform of _f_(_r_) cos<sup>_n_</sup> θ is _F<sub>_n_</sub>_(_r_) cos<sup>_n_</sup> θ, with some other radial part _F_<sub>_n_</sub> (different for different powers _n_) but the same angular dependence. And if a suitable basis is chosen for the radial part, the forward transforms _f_(_r_) → _F_<sub>_n_</sub>(_r_) can be computed analytically for all _n_. The rest is as usual in all “BASEX”-type methods.

That is, to perform a forward or inverse Abel transform of an image, it is sufficient to extract from it the radial profiles of all cos<sup>_n_</sup> θ terms, transform these radial profiles (which is much faster that transforming the whole image) and, if needed, construct the transformed image from the transformed profiles. The transformed radial profiles also directly give various radial distributions needed in VMI studies, thus if the transformed image itself is not needed, the last step can be skipped, making the analysis even faster.

An example of speed comparison with `two_point`, the fastest of current methods:
```
PyAbel benchmark run on AMD64 Family 23 Model 17 Stepping 0, AuthenticAMD

time in milliseconds

=== Basis generation ======================================================

Method              n = 513    n = 1025    n = 2049    n = 4097    n = 8193
---------------------------------------------------------------------------
rbasex                 11.7        34.8       116.7       598.9      3248.7
two_point               2.6        13.7        57.7       234.6      1019.7


=== Forward Abel transform ================================================

Method              n = 513    n = 1025    n = 2049    n = 4097    n = 8193
---------------------------------------------------------------------------
rbasex                  2.7        14.6        62.7       302.2      1167.8
rbasex(None)            1.4         6.5        31.7       132.7       516.2


=== Inverse Abel transform ================================================

Method              n = 513    n = 1025    n = 2049    n = 4097    n = 8193
---------------------------------------------------------------------------
rbasex                  2.7        14.7        63.0       302.6      1170.6
rbasex(None)            1.4         6.6        31.8       132.8       516.8
two_point               1.3         9.9        73.7       561.5      4343.1
```
(here `rbasex(None)` means that the output image is not constructed)
And, for completeness, the time that must be added to the `two_point` transform to extract radial distributions from the image:
```
order = 2, time in milliseconds
=== Full image, cached ======================================================
Method                n = 513    n = 1025    n = 2049    n = 4097    n = 8193
-----------------------------------------------------------------------------
linear
   MIN, none              1.8         7.6        31.0       125.7       498.1
```
So when only the radial distributions are needed, the rBasex transform is almost free. Otherwise, when the transformed image is also constructed, `rbasex` is competitive with `two_point` (and other fast methods) at 1 megapixel (the typical image size) and wins at larger image sizes. The best thing is that the time complexity scales only quadratically with the image radius, in contrast to the cubic scaling in all other methods (except maybe Hansen–Law).

Now let's see how the documentation will be built for this PR...

(More testing, proofreading and fact-checking about pBasex are still neded.)

MikhailRyazanov commented:
Addendum: some small changes to the `abel.tools.vmi.Distributions` class are also included. Some are directly related to `rbasex`, and some are just usability/consistency improvements.

stggh commented:
This is neat!  I will examine the method for some VMI images that have an ion-beam (MCP depleted gain) hole.

DanHickstein commented:
This is super rad! I will proofread the code more when time allows.

For now, I'm going to close and re-open this PR to see if I can trigger the readthedocs check.

MikhailRyazanov commented:
@DanHickstein, great! Now the docs are available.

> I will proofread the code more when time allows.

Proofread the _documentation_, please! :–) Grammar, awkward wording and if something isn't clear...

In the meantime I'll think about adding more unit tests (basis, caching, image shape and that higher orders work correctly).

MikhailRyazanov commented:
@DanHickstein and @stggh, if you have time, please review this PR...

stggh commented:
Away 10/3-20/3 in New Zealand (North island). The trip was originally planned for Hokkaido (Japan), but cancelled due to the State of Emergency there. 

I have been running some of my VMI data that have "holes" due to MCP ion-beam gain depletion, but haven't yet come across an example that demonstrates the power of your method, or caused a problem. Your code and methods are very good.

I will review on my return, but otherwise, happy to accept this PR if you feel the delay is too long.

MikhailRyazanov commented:
@stggh, this is not urgent, so looking at it after your return is fine.

The ability to analyze data with “holes” is just a bonus; you'll probably need very poor images (with a hole disturbing a significant angular range) to see a clear difference in real data (but it would be interesting if you can find an example!). The main benefit of this method is that it produces clean images, like lin-BASEX, but actually even cleaner and works faster. Plus, it is a high-resolution substitution for pBasex for those who want it.

MikhailRyazanov commented:
ping

stggh commented:
Thanks for the reminder! 

I have been caught up with 2 weeks home isolation, following our return from NZ, which might appear ideal for providing time for a code review, except I caught a flu. The associated COVID-19 test was negative. 

I like your excellent synopsis:
> The main benefit of this method is that it produces clean images, like lin-BASEX, but actually even cleaner and works faster. Plus, it is a high-resolution substitution for pBasex for those who want it.

I will get on to it!

DanHickstein commented:
This looks awesome Mikhail! Sorry to take so long to review it. I should have some time over the next few days.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+
+        In all cases, `strength` = 0 provides no regularization. For the
+        Tikhonov methods, `strength` ~ 100 is a reasonable value for megapixel
+        images. For truncated SVD, `strength` must be < 1; `strength` ~ 0.1 is
```
For the SVD method, is the effect of the `strength` independent of image size? For the Tikhonov methods, should we try to find a way to scale the strength of the regularization so that the effect is independent of image size? It might be something simply like simply multiplying the strength by the image size...  It would be nice if we could just recommend a strength and it would more-or-less work for all images.

MikhailRyazanov replied:
I don't actually know. For SVD, I use a fraction instead of an absolute number, since it, in theory, should be less size-dependent. But it still has somewhat different effects for different image sizes. The “L2” method essentially folows the BASEX approach, and its `strength` actually gives effects on velocity distributions very similar to that of BASEX `reg`. So:
1. This question of the optimal regularization strength needs deeper research. I have seen some attempts in the literature, but no definitive answer.
2. So if we want to give more specific recommendations, they perhaps deserve a separate documentation section (with references, examples and so on).
3. I actually have some examples in `transform_methods/rbasex-math.html#examples`, but if you take `rbasex-regRMSE.py` and try it with different images and even different random samples of the _same_ image, you'll see that the optimal parameters will be different. (Hence there is a big “Warning” preceding the examples.)
3. If we want to introduce some automatic scaling to reduce the size dependence, it probably should be done for BASEX as well, for consistency.

MikhailRyazanov replied:
Added a hyperlink to the end of that paragraph.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+        ``'same'`` (default):
+            same shape and origin as the input
+        ``'fold'`` (fastest):
+            Q0 quadrant (for ``odd=False``) or right half (for ``odd=True``) up
```
change "Q0" to "Q0 (upper left)"?

MikhailRyazanov replied:
Upper right actually.

MikhailRyazanov replied:
Done.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+            to **rmax**, but limited to the largest input-image quadrant (or
+            half)
+        ``'unfold'``:
+            like ``'fold'``, but symmetrically “unfolded”
```
I'm not really sure what this means. Basically, you calculate just Q0 and then duplicate it to all of the other quadrants? Isn't this what the transform always does? Or does it normally allow for asymmetry in the top/down and left/right dimensions?

MikhailRyazanov replied:
Like lin-BASEX, it assumes that the image is composed of spherical harmonics, so for even-only orders it is totally symmetric, for even-only orders is horizontally symmetric and vertically antisymmetric, and for mixed orders is just horizontally symmetric. In principle, you can transform each quadrant independently, but this will yield 4 different velocity distributions. Averaging them will give the same results as transforming the whole image, so this is not very useful.
The transform itself actually produces only the velocity distribution. The transformed image is then created from that distribution in the form requested (internally, yes, just Q0 of right half is calculated and, if needed, is mirrored to other parts).

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    origin : tuple of int or str
+        image origin, explicit in the (row, column) format, or as a location
+        string
+    rmax : int or string
```
What string values can be entered? Case sensitive?

MikhailRyazanov replied:
There is a paragraph “The origin, rmax, order, odd and weights parameters are passed to abel.tools.vmi.Distributions, so see its documentation for their detailed descriptions.” (with a hyperlink) just above the “Parameters” section. :–)

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+        image origin, explicit in the (row, column) format, or as a location
+        string
+    rmax : int or string
+        largest radius to include in the transform
```
Say what the default value is.

MikhailRyazanov replied:
It is also described in `abel.tools.vmi.Distributions`. Do you want to include it here as well? (I guess, something like “by default, the largest radius fitting in at least one full quadrant”.)

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    rmax : int or string
+        largest radius to include in the transform
+    order : int
+        highest angular order present in the data, ≥ 0
```
Say what the default value is.

MikhailRyazanov replied:
Done

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    order : int
+        highest angular order present in the data, ≥ 0
+    odd : bool
+        include odd angular orders (enabled automatically if **order** is odd)
```
Default is False.

MikhailRyazanov replied:
Done.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    weights : m × n numpy array, optional
+        weighting factors for each pixel. The array shape must match the image
+        shape. Parts of the image can be excluded from analysis by assigning
+        zero weights to their pixels.
```
"The default is None, which applies equal weight to all pixels."

MikhailRyazanov replied:
Done.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+        shape. Parts of the image can be excluded from analysis by assigning
+        zero weights to their pixels.
+    direction : str: ``'forward'`` or ``'inverse'``
+        type of Abel transform to be performed
```
default is 'inverse'

MikhailRyazanov replied:
Done.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    direction : str: ``'forward'`` or ``'inverse'``
+        type of Abel transform to be performed
+    reg : None or str or tuple (str, float), optional
+        regularization to use for inverse Abel transform. ``None`` means no
```
"default is none"

stggh replied:
`reg='L2'` i.e. string only fails
```python
    transform_options=dict(reg='L2'))
       :
   ValueError: Wrong regularization format "L2"
```

MikhailRyazanov replied:
I have changed `'L2'` to `('L2', strength)` and so on. This should be more clear, I guess.

MikhailRyazanov replied:
I don't know. I think that it is better to keep the API documentation and the method description separated as much as possible. @stggh also said once that he reads these docstrings within the interactive interpreter session, so it's better to keep them readable as plain text and, I guess, concise. Plus, I hope that people read “How it works/When to use it/How to use it” instead of randomly playing with the parameters (or at least before that). So it is probably better to add more information there (this applies to other methods as well; for example, our BASEX documentation tells almost nothing about regularization).

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+        path to the directory for saving / loading the basis set (useful only
+        for the inverse transform without regularization; time savings in other
+        cases are small and might be negated by the disk-access overhead).
+        If ``None``, the basis set will not be loaded from or saved to disk.
```
Default is None.

MikhailRyazanov replied:
Done.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+        cases are small and might be negated by the disk-access overhead).
+        If ``None``, the basis set will not be loaded from or saved to disk.
+    verbose : bool
+        print information about processing (for debugging)
```
default is False.

MikhailRyazanov replied:
Done.

DanHickstein commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    recon : 2D numpy array or None
+        the transformed image. Is centered and might have different dimensions
+        than the input image.
+    distr : Distributions.Results object
```
This is fundamentally different from all of the other `_transform` functions, which only return the image array. I see why it needs to work this way, but it's potentially confusing.

stggh replied:
`distr` is encapsulates `(r)basex` results, much like the `Transform` class contains more than just the Abel transformed image. It is neat and tidy, and e.g. should be used for `linbasex` to contain the `Beta` arrays unique to it.

MikhailRyazanov replied:
Yes, this is very much like in `linbasex`, except that `rbasex` always returns 2 values instead of 1 or 3 depending on input parameters.

DanHickstein commented:
The comparison with the other methods looks nice! 

![image](https://user-images.githubusercontent.com/1107796/78102857-96cdef80-73a8-11ea-9b77-ba2066332191.png)

https://gist.github.com/DanHickstein/5874a3bee2dc0eb37cc724049937fcc4

DanHickstein commented:
Wow. I am more and more impressed with this work. I was going to ask if you were going to write a paper about it, but it looks like you've already written one! :)  https://pyabel--270.org.readthedocs.build/en/270/transform_methods/rbasex-math.html#rbasexmath

MikhailRyazanov commented:
Thanks, Dan! I'll wait for you to finish your review (please also make any corrections to the text!) and for Steve's review (@stggh, it's sad that you've got a flu, but hopefully it is less dangerous than this new coronavirus. Get well soon!), and then I'm going to make the edits.

> I was going to ask if you were going to write a paper about it, but it looks like you've already written one! :)

Well, this was a practice for a real one, which I hope to publish eventually (about the original method and this faster version). When I get time to work on it...

stggh commented:
Amazing and very impressive work @MikhailRyazanov!  

The only issue that has appeared in my analysis is a factor of <math>&pi;</math> between the radial intensity yielded by `.distr.rIbeta()` and that from `.angular_integration`, no doubt a fault  with the latter :-(

```python
im = np.loadtxt('O2-ANU1024.txt.bz2')
imc = abel.tools.center.center_image(im, method='slice')

rbasex = abel.Transform(imc, method='rbasex', angular_integration=True)

rb, rp, rp1 = rbasex.distr.rIbeta()
radial, pes = rbasex.angular_integration
```
Here is the O<sub>2</sub><sup>-</sup> PES, that includes the factor <math>1/&pi;</math> for the `distr` intensity (blue line).
<img width="596" alt="Screen Shot 2020-04-03 at 10 51 05" src="https://user-images.githubusercontent.com/10932229/78310263-17534280-7599-11ea-8fa1-575670406ba7.png">
The 'extra' intensity is due to the better resolution of the basis decomposition.
<img width="578" alt="Screen Shot 2020-04-03 at 10 58 03" src="https://user-images.githubusercontent.com/10932229/78310577-17077700-759a-11ea-9420-299623479f6b.png">

The same &pi; scaling factor applies for other species, e.g. C<sub>2</sub>H<sup>-</sup>:
<img width="584" alt="Screen Shot 2020-04-03 at 10 55 34" src="https://user-images.githubusercontent.com/10932229/78310454-c2fc9280-7599-11ea-8661-e3af2355746b.png">

stggh commented on abel/benchmark.py:
```diff
@@ -9,9 +9,10 @@
 from . import hansenlaw
+from . import linbasex
 from . import onion_bordas
+from . import rbasex
```
`rbasex` has a nice fast execution speed!

stggh commented on examples/example_rbasex_multimass.py:
```diff
@@ -0,0 +1,131 @@
+plt.autoscale(enable=True, tight=True)
+
+plt.tight_layout()
+plt.show()
```
Impressive!

MikhailRyazanov replied:
:–)

MikhailRyazanov commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    IM : m × n numpy array
+        the image to be transformed
+    origin : tuple of int or str
+        image origin, explicit in the (row, column) format, or as a location
```
There is a paragraph “The origin, rmax, order, odd and weights parameters are passed to abel.tools.vmi.Distributions, so see its documentation for their detailed descriptions.” (with a hyperlink) just above the “Parameters” section. :–)

MikhailRyazanov commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    ----------
+    IM : m × n numpy array
+        the image to be transformed
+    origin : tuple of int or str
```
OK.

MikhailRyazanov commented on abel/rbasex.py:
```diff
@@ -0,0 +1,751 @@
+    reg : None or str or tuple (str, float), optional
+        regularization to use for inverse Abel transform. ``None`` means no
+        regularization, a string selects a non-parameterized regularization
+        method, and parameterized methods are selected by a tuple (`method`,
```
I have changed 'L2' to ('L2', strength) and so on. This should be more clear, I guess.

MikhailRyazanov replied:
The [new code](https://github.com/MikhailRyazanov/PyAbel/blob/bd456a171e9810030be2bec8c4454cea57c3352d/abel/rbasex.py#L89-L110) and [how it renders](https://pyabel--270.org.readthedocs.build/en/270/abel.html#abel.rbasex.rbasex_transform).

MikhailRyazanov commented:
@stggh, we have discussed the π factor in #257. Basically, `rbasex` is based on `Distributions`, so it has all the same properties. My _I_(_r_) follows the usual convention that 4π ∫ _I_(_r_) _r_<sup>2</sup>_dr_ equals the total intensity (the number of events). I don't remember what exactly `angular_integration` calculates and why. But at least the reconstructed images are consistent with other methods, so you can use either `Distributions` or `angular_integration` in any case. :–)

Can you please do C<sub>2</sub>H<sup>−</sup> with `reg='pos'` and show the low-intensity portion of that plot (say, up to 0.2e5 or 0.1e5)?

stggh commented:
> #257 

 Oh, that is right, thanks for the reminder.

> you can use either `Distributions` or `angular_integration` in any case

 Yes, this works very well.

> C<sub>2</sub>H<sup>−</sup> with reg='pos' 

Note in this example`reg='pos'` affects both `Distributions` and `angular_integration`. `reg` produces some nice smoothing.

<img width="590" alt="Screen Shot 2020-04-05 at 09 24 25" src="https://user-images.githubusercontent.com/10932229/78463385-4e536080-771f-11ea-90f1-44f33a8b6a42.png">

<img width="588" alt="Screen Shot 2020-04-05 at 09 25 05" src="https://user-images.githubusercontent.com/10932229/78463390-5d3a1300-771f-11ea-9294-48bc61d80572.png">

I will send you the (unpublished) VMI image data via Cloudstor, in case you want to play without the delay of waiting for me to run your nice `rbasex` analysis.

MikhailRyazanov commented:
Thanks for sending the data! Now I see that the negative dip at small radii is due to a dark spot near the image center:
![C2H-center](https://user-images.githubusercontent.com/12854133/78512130-5cc16c80-775f-11ea-80d4-bdb0663c08d1.png)
If I mask it, together with the strange contamination in the lower left quadrant, and the strange vertical line (just in case), as described [here](https://pyabel--270.org.readthedocs.build/en/270/transform_methods/rbasex.html#how-to-use-it):
![C2H-mask](https://user-images.githubusercontent.com/12854133/78512219-e6713a00-775f-11ea-9f79-22a78b8804bd.png)
I get more reasonable intensity behavior near the center in both cases — no significant negative intensities beyond the noise level without regularization, and no suspicious hard clamping to zero with the positive regularization:
![C2H-plot](https://user-images.githubusercontent.com/12854133/78512230-fb4dcd80-775f-11ea-8fbd-30f725b343f2.png)

In any case, it is interesting to see (more obviously in your plots) how the positivity constraints clean up the 300–360 region, such that the peak near 320 becomes apparent. Do you know whether it is real?

stggh commented:
Excellent! Thanks for taking a look at the data. The centre line is due to the two CMOS sensors of the camera. The MCP gain depletion results from the ion-beam, in our coaxial spectrometer, which is present during set-up, but switched out, when collecting photoelectrons, although, the gain depletion remains.

You make some good observations, thanks to the (r)basex regularization analysis.

Off-hand I cannot easily say anything about the possibility of a transition near 320 radius. There are many transitions in this region, and it is heavily perturbed. However, I will take a look, in relation to the assignments that have been made. It may take a little time.

DanHickstein commented on doc/transform_methods/rbasex.rst:
```diff
@@ -0,0 +1,211 @@
+   functions. This makes the projected basis functions also separable into
+   radial and angular parts.
+
+3. The basis separability allows decomposition of the problem in two steps:
```
I would change the first part to read: "first, radial distributions are extracted without
   intermediate rebinning to polar grid (which is faster and avoids accumulation
   of resampling errors)"

DanHickstein replied:
Most important is to fix the "what" -> "which"

DanHickstein commented on doc/transform_methods/rbasex.rst:
```diff
@@ -0,0 +1,211 @@
+When to use it
+--------------
+
+Although this method can be applied only to velocity-map images, and not just
```
There might be other situations where the data looks close enough to a VMI image so that rBasex might be appropriate. So, I'd suggest that we soften the language to something like:

"The rBasex method makes additional assumptions about the data (beyond cylindrical symmetry), so is most applicable for the transform of velocity-map images. In this special case, if offers several benefits:"

DanHickstein commented on doc/transform_methods/rbasex.rst:
```diff
@@ -0,0 +1,211 @@
+Although this method can be applied only to velocity-map images, and not just
+cylindrically symmetric images, it has several benefits in this special case.
+
+1. The reconstructed radial distributions, which are of the primary interest in
```
"which are *often* the primary interest".

MikhailRyazanov commented on doc/transform_methods/rbasex.rst:
```diff
@@ -0,0 +1,211 @@
+
+This method resembles the pBasex [1]_ approach of expanding a velocity-map
+image over a 2D basis set in polar coordinates, but uses more convenient basis
+functions with analytical Abel transforms, developed by M. Ryazanov in [2]_.
```
OK, removing “in” should indeed work.

MikhailRyazanov commented on doc/transform_methods/rbasex.rst:
```diff
@@ -0,0 +1,211 @@
+an :math:`N \times N` image has only :math:`N_r \times N_a` degrees of freedom,
+where :math:`N_r` is the number of radial samples, usually :math:`N / 2`, and
+:math:`N_a` is the number of angular terms, a small number depending on the
+number of photons. These are the “radial distribution” extracted from the
```
As far as I know, sometimes the involved orbitals and other factors can lower this number to some effective order < the number of photons, but this is not always known in advance. Can this be worded better without going too far?

MikhailRyazanov replied:
In the laboratory frame the photoelectron angular distribution is directly related to the orbital angular momentum carried out by the electron, which must be equal to the angular momentum change upon the transition, which is in turn limited by the number of absorbed photons (each gives 0 or ±1). This is true for any non-oriented systems, as far as I know (from reading Zare's works). For oriented systems this is more complicated and generally involves infinitely many spherical harmonics. @stggh should know more...

MikhailRyazanov commented on doc/transform_methods/rbasex.rst:
```diff
@@ -0,0 +1,211 @@
+Abel transforms of the basis functions can be calculated easily if the basis is
+chosen wisely.
+
+See :ref:`rBasexmath` for the complete description.
```
Well, if you click on [+] to expand “How it works” in the TOC, it will show up there. :–) I'll try to check whether it can be expanded there by default. About including everything in the main document, I don thick that it is a good idea to just put everything in “How it works”. Maybe insert it before “Citation"? But is is really long and heavy.

MikhailRyazanov commented on doc/transform_methods/rbasex-coord.py:
```diff
@@ -0,0 +1,92 @@
+an = 11  # angle arc segments
+anc = an // 2  # index of central point
+
+fig = plt.figure(figsize=(3, 3)) #??, frameon=False
```
I think, by default Matplotlib was adding frames around figures, which I do not want, so I had `frameon=False` everywhere. But this one did not have a frame even without this option. I should better include it here also for consistency.

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+.. plot:: transform_methods/rbasex-coord.py
+    :align: center
+
+This figure also illustrated important relations between the 3D and 2D radii:
```
Thanks!

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+    \end{cases}
+
+(and :math:`b_0(\rho)` does not have the inner part, :math:`R - 1 < \rho < R`,
+since :math:`\rho \geqslant 0`). These functions actually form a basis of the
```
What about “These functions form a basis of piecewise linear functions with nodes at each pixel, so for any radial distribution, its piecewise linear approximation on the pixel grid can be represented in this basis.”?

DanHickstein replied:
Maybe offer the layman's explanation afterwards: "In other words, this basis set allows us to draw straight lines between pixel values, which is the simplest way to interpret a pixel-based dataset." Or something like this.

MikhailRyazanov commented on doc/transform_methods/rbasex-peak.py:
```diff
@@ -0,0 +1,73 @@
+         [ 0,     0,   0.4, 0.6,   0.4,     0,        0],
+         'go--', label='3 px')
+P = 0.4 * p(R - 1, n) + 0.6 * p(R, n) + 0.4 * p(R + 1, n)
+plt.plot(r, P, 'g')
```
OK, changed to cyan.

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+:math:`\mathbf P_n^{\rm T}` are upper triangular, and their inversions
+:math:`\big(\mathbf P_n^{\rm T}\big)^{-1}` are also upper triangular for all
+:math:`n`, which additionally facilitates the computations. (This triangularity
+makes the inverse Abel transform similar to the “onion peeling” procedure
```
Maybe, but which? If I refer to `onion_bordas`, then I need to explain that here I “peel” spherical shells instead of circular rings in cylindrical coordinates. And then it alludes to the “polar onion-peeling” (http://dx.doi.org/10.1063/1.3126527), but that method is actually not correct (their equation 6 is valid only for infinitely narrow radial function), although it works more or less fine at sufficiently large radii.

DanHickstein replied:
Well, which "onion peeling" procedure did you have in mind when you wrote this? Just the general idea of working from the outside in?

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+Abel transform, step 3 is not needed at all, and for the inverse Abel
+transform, its results can be cached. Thus processing a sequence of images
+takes time linearly proportional to the total number of processed pixels. In
+other words, the throughput is independent on the image size.
```
By the “throughput” I mean the number pixels per second, as we define it in RSI. The number of images per second is the “frame rate”. I can write “pixel throughput” to make it unambiguous.

DanHickstein replied:
Okay, I understand,. Maybe write: "In other words, the throughput (measured in pixels per second) is independent on the image size."

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+   time.
+3. For the inverse Abel transform, the :math:`\mathbf P_n^{\rm T}` matrices are
+   inverted, in :math:`O(N\,R_\text{max}^3)` total time. This step is not
+   needed for the forward Abel transform.
```
Not really “of the same image size”, but most other parameters also must be the same. I discuss caching immediately after describing these steps. Do you think that it should be mentioned in this mathematical part also?

DanHickstein replied:
Good point. Probably the discussion below is sufficient.

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+
+from a quadratic minimization (“least-squares”) problem, which is equivalent to
+the original problem, but makes evident that for ill-conditioned problems the
+minimum is very “flat”, and many different :math:`\mathbf x` can be accepted as
```
Well, **x** is a vector. I think, “values of **x**” is formally correct, but sounds misleading. Maybe “...many different vectors **x** can be accepted as a solution”?

DanHickstein replied:
that wording also works for me.

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+Truncated SVD
+^^^^^^^^^^^^^
+
+This is the method mentioned in the pBasex article (but not actually used in
```
Done.

MikhailRyazanov commented on doc/transform_methods/rbasex-sim.py:
```diff
@@ -0,0 +1,53 @@
+# simulated projection fith Poissonian noise
+proj, _ = rbasex_transform(source, direction='forward')
+proj[proj < 0] = 0
+proj = np.random.RandomState(0).poisson(proj)  # (reproducible, see NEP 19)
```
NumPy Enhancement Proposal 19 — Random Number Generator Policy:
https://numpy.org/neps/nep-0019-rng-policy.html
This link is probably stable, so I can include it instead of “NEP 19”.

DanHickstein replied:
Cool!

MikhailRyazanov commented on doc/transform_methods/rbasex-regRMSE.py:
```diff
@@ -0,0 +1,83 @@
+
+# finish
+plt.tight_layout()
+
```
Done.

MikhailRyazanov commented on doc/transform_methods/rbasex-math.rst:
```diff
@@ -0,0 +1,810 @@
+
+
+Using ``reg=None``
+""""""""""""""""""
```
This will be difficult. The tick labels already overlap with the image to save space. This is not very noticeable, but the vertical-axis label will be drawn over the image and look very bad. Making the panels separated will squeeze the plots too much, and the whole figure cannot be made wider. Maybe this is not a big deal, since everything is explained immediately below the figure?

DanHickstein replied:
okay, we can skip the labels.

MikhailRyazanov commented:
Thanks, @DanHickstein! I have made some changes and asked about less trivial ones, please take a look.

DanHickstein commented:
Okay, I went through and replied to your questions and comments. Please let me know if I missed any. It looks like there are still some changes that you still need to make.

MikhailRyazanov commented:
@DanHickstein, I have more or less implemented your suggestions: restructured “How to use it” to start from `Transform`; added some words (and a reference) about regularization to “When to use it”; added a reference to `onion_peeling` with some more details; plus did other, minor corrections.

Most importantly, I have added a check to `Transform` that raises an error if its `origin` is used with `method='rbasex'` in a wrong way. I think, it is better to fail and show a useful message instead of producing unexpected results (with or without a warning).

If everything looks _fine_, maybe let's merge this? The few remaining issues, like restructuring the documentation, can be addressed later (I'm planning to do some other additions to the method anyway).

DanHickstein commented:
@MikhailRyazanov, I had a look at the most recent changes that you've made and they look great!

I think that having Transform raise an error is the right way to go. The error messages are descriptive and should be quite helpful for users. 

I think that you can go ahead and merge. Thank you for this fantastic addition!

MikhailRyazanov commented:
Thanks!
Merging...

DanHickstein commented:
Great work!
